core/src/main/java/org/elasticsearch/cluster/routing/DelayedAllocationService.java

```diff
@@ -178,8 +178,8 @@ private void removeIfSameTask(DelayedRerouteTask expectedTask) {
     /**
      * Figure out if an existing scheduled reroute is good enough or whether we need to cancel and reschedule.
      */
-    private void scheduleIfNeeded(long currentNanoTime, ClusterState state) {
-        assertClusterStateThread();
+    private synchronized void scheduleIfNeeded(long currentNanoTime, ClusterState state) {
+        assertClusterOrMasterStateThread();
```

Comment:
The service is a `ClusterStateListener` registered on the `ClusterApplierService` to react to `clusterChanged` events. Logically it would make sense to register the listener on the `MasterService` (as `DelayedAllocationService` is purely master-driven), but `MasterService` currently does not provide such listener functionality. Instead of adding that infrastructure to `MasterService`, I preferred to synchronize this method for now and wait for other usage scenarios to pop up that warrant changes in `MasterService`.